backend/elasticsearch-client/elasticsearch-java-client/src/main/java/org/hibernate/search/backend/elasticsearch/client/java/impl/ClientJavaHttpRequestInterceptor.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.search.backend.elasticsearch.client.java.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.search.backend.elasticsearch.client.common.spi.ElasticsearchRequestInterceptor;
+import org.hibernate.search.backend.elasticsearch.client.common.spi.ElasticsearchRequestInterceptorContext;
+import org.hibernate.search.util.common.AssertionFailure;
+
+import org.apache.hc.client5.http.protocol.HttpClientContext;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.core5.http.HttpEntityContainer;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpRequestInterceptor;
+import org.apache.hc.core5.http.NameValuePair;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.hc.core5.net.URIBuilder;
+
+record ClientJavaHttpRequestInterceptor(ElasticsearchRequestInterceptor elasticsearchRequestInterceptor)
+		implements HttpRequestInterceptor {
+
+	@Override
+	public void process(HttpRequest request, EntityDetails entity, HttpContext context) throws IOException {
+		elasticsearchRequestInterceptor.intercept(
+				new ClientJavaRequestContext( request, entity, context )
+		);
+	}
+
+	private record ClientJavaRequestContext(HttpRequest request, EntityDetails entity, HttpClientContext clientContext)
+			implements ElasticsearchRequestInterceptorContext {
+		private ClientJavaRequestContext(HttpRequest request, EntityDetails entity, HttpContext context) {
+			this( request, entity, HttpClientContext.cast( context ) );
+		}
+
+		@Override
+		public boolean hasContent() {
+			return entity != null;
+		}
+
+		@Override
+		public InputStream content() throws IOException {
+			HttpEntity localEntity = null;
+			if ( entity instanceof HttpEntity httpEntity ) {
+				localEntity = httpEntity;
+			}
+			else if ( request instanceof HttpEntityContainer entityContainer ) {
+				localEntity = entityContainer.getEntity();
+			}
+
+			if ( localEntity != null ) {
+				if ( !localEntity.isRepeatable() ) {
+					throw new AssertionFailure( "Cannot sign AWS requests with non-repeatable entities" );
+				}
+				return localEntity.getContent();
+			}
+
+			if ( entity instanceof AsyncEntityProducer producer ) {
+				if ( !producer.isRepeatable() ) {
+					throw new AssertionFailure( "Cannot sign AWS requests with non-repeatable entities" );
+				}
+				return new HttpAsyncEntityProducerInputStream( producer, 1024 );
+			}
```

Comment:
This one is a bit annoying as there seems to be no easier/simpler way to get the entity out of the request, but well ... it is what it is 🙂 🤷🏻

Reply:
IIUC you copy-pasted the old `HttpAsyncContentProducerInputStream` two times where you needed it? And possibly other related classes?

This makes me uneasy, because this is very fragile code, and my memories of working with it are... not pleasant ones.

Would it make sense to abstract over the various HTTPD classes (content encoder, ...) and move this code to a single implementation in `-utils` instead?